pkg/depthmonitor/depthmonitor.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package depthmonitor
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+const (
+	depthKey             string  = "storage_depth"
+	adaptationFullWindow float64 = 2 * 60 * 60 // seconds allowed to fill half of the fully empty reserve
+	adaptationRollback           = 5 * 60      // seconds to slightly roll back the adaption window in case half capacity is not reached
+)
+
+var (
+	manageWait = 5 * time.Minute
+)
+
+// ReserveReporter interface defines the functionality required from the local storage
+// of the node to report information about the reserve. The reserve storage is the storage
+// pledged by the node to the network.
+type ReserveReporter interface {
+	// Current size of the reserve.
+	Size() (uint64, error)
+	// Capacity of the reserve that is configured.
+	Capacity() uint64
+}
+
+// SyncReporter interface needs to be implemented by the syncing component of the node (pullsync).
+type SyncReporter interface {
+	// Rate of syncing in terms of chunks/sec.
+	Rate() float64
+}
+
+// Topology interface encapsulates the functionality required by the topology component
+// of the node.
+type Topology interface {
+	topology.NeighborhoodDepther
+	topology.SetStorageDepther
+}
+
+// Service implements the depthmonitor service
+type Service struct {
+	depthLock        sync.Mutex
+	topology         Topology
+	syncer           SyncReporter
+	reserve          ReserveReporter
+	st               storage.StateStorer
+	logger           logging.Logger
+	quit             chan struct{}
+	wg               sync.WaitGroup
+	storageDepth     uint8
+	oldStorageRadius uint8
+}
+
+// New constructs a new depthmonitor service
+func New(
+	t Topology,
+	syncer SyncReporter,
+	reserve ReserveReporter,
+	st storage.StateStorer,
+	logger logging.Logger,
+	warmupTime time.Duration,
+) *Service {
+
+	s := &Service{
+		topology: t,
+		syncer:   syncer,
+		reserve:  reserve,
+		st:       st,
+		logger:   logger,
+		quit:     make(chan struct{}),
+	}
+
+	s.wg.Add(1)
+	go s.manage(warmupTime)
+
+	return s
+}
+
+func (s *Service) manage(warmupTime time.Duration) {
+	defer s.wg.Done()
+
+	// wait for warmup
+	select {
+	case <-s.quit:
+		return
+	case <-time.After(warmupTime):
+	}
+
+	// check if we have already saved a depth value before shutting down previously.
+	var initialDepth uint8
+	err := s.st.Get(depthKey, &initialDepth)
+	if err != nil && !errors.Is(err, storage.ErrNotFound) {
+		s.logger.Errorf("depthmonitor: failed reading stored depth: %s", err.Error())
+	}
+
+	// if we are starting from scratch, we can use the initial NeighborhoodDepth as a starting point.
+	if initialDepth == 0 {
+		initialDepth = s.topology.NeighborhoodDepth()
+	}
```

Comment:
If `initialDepth` is a stored depth, we should call `topology.SetStorageDepth` with it in the start as the manage worker might never kick-off.